`chezmoi dump` crashes

## Describe the bug

After having read about `chezmoi dump` in the documentation I wanted to play around with it. However for me it just crashes:

```console
$ chezmoi dump            
chezmoi: warning: config file template has changed, run chezmoi init to regenerate config file
panic: update to no update system

goroutine 1 [running]:
github.com/twpayne/chezmoi/v2/pkg/chezmoi.noUpdateSystemMixin.RunCmd(...)
	/home/runner/work/chezmoi/chezmoi/pkg/chezmoi/system.go:84
github.com/twpayne/chezmoi/v2/pkg/chezmoi.(*TargetStateModifyDirWithCmd).Apply(0x1de21e0, {0xf782d0, 0x1f93df8}, {0xf76ff4, 0x1f93de0}, {0xf7495c, 0x1e28b88})
	/home/runner/work/chezmoi/chezmoi/pkg/chezmoi/targetstateentry.go:81 +0x104
github.com/twpayne/chezmoi/v2/pkg/chezmoi.(*SourceState).Apply(0x1f72180, {0xf782d0, 0x1f93df8}, {0xf78460, 0x1f93de8}, {0xf76ff4, 0x1f93de0}, {{0x0, 0x0}}, {{0x1df44b0, ...}}, ...)
	/home/runner/work/chezmoi/chezmoi/pkg/chezmoi/sourcestate.go:681 +0xd5c
github.com/twpayne/chezmoi/v2/pkg/cmd.(*Config).applyArgs(0x1e44000, {0xf757ac, 0x1c260e0}, {0xf782d0, 0x1f93df8}, {{0x0, 0x0}}, {0x15603c4, 0x0, 0x0}, ...)
	/home/runner/work/chezmoi/chezmoi/pkg/cmd/config.go:619 +0x734
github.com/twpayne/chezmoi/v2/pkg/cmd.(*Config).runDumpCmd(0x1e44000, 0x2266160, {0x15603c4, 0x0, 0x0})
	/home/runner/work/chezmoi/chezmoi/pkg/cmd/dumpcmd.go:43 +0x118
github.com/spf13/cobra.(*Command).execute(0x2266160, {0x15603c4, 0x0, 0x0})
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.4.0/command.go:856 +0x6b0
github.com/spf13/cobra.(*Command).ExecuteC(0x1c24580)
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.4.0/command.go:974 +0x414
github.com/spf13/cobra.(*Command).Execute(...)
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.4.0/command.go:902
github.com/twpayne/chezmoi/v2/pkg/cmd.(*Config).execute(0x1e44000, {0x1c0c048, 0x1, 0x1})
	/home/runner/work/chezmoi/chezmoi/pkg/cmd/config.go:1120 +0x78
github.com/twpayne/chezmoi/v2/pkg/cmd.runMain({{0xf6f338, 0x6}, {0xf76bb8, 0x28}, {0xf72d78, 0x14}, {0xf708d8, 0xa}}, {0x1c0c048, 0x1, ...})
	/home/runner/work/chezmoi/chezmoi/pkg/cmd/cmd.go:264 +0x148
github.com/twpayne/chezmoi/v2/pkg/cmd.Main({{0xf6f338, 0x6}, {0xf76bb8, 0x28}, {0xf72d78, 0x14}, {0xf708d8, 0xa}}, {0x1c0c048, 0x1, ...})
	/home/runner/work/chezmoi/chezmoi/pkg/cmd/cmd.go:128 +0x38
main.main()
	/home/runner/work/chezmoi/chezmoi/main.go:26 +0xd8
```

**NOTE!** Only dump says "chezmoi: warning: config file template has changed, run chezmoi init to regenerate config file". Running `chezmoi init` does not remove this warning. So best as I can tell that line is a lie.

This happens both on my Pi (official binary build) and on my desktop (Arch Linux package).

## To reproduce

```console
$ chezmoi dump
```

## Expected behavior

Not crash?

## Output of command with the `--verbose` flag

Exactly the same:

```console
$ chezmoi dump --verbose
chezmoi: warning: config file template has changed, run chezmoi init to regenerate config file
panic: update to no update system

goroutine 1 [running]:
github.com/twpayne/chezmoi/v2/pkg/chezmoi.noUpdateSystemMixin.RunCmd(...)
	/home/runner/work/chezmoi/chezmoi/pkg/chezmoi/system.go:84
github.com/twpayne/chezmoi/v2/pkg/chezmoi.(*TargetStateModifyDirWithCmd).Apply(0x21e7c50, {0xf782d0, 0x2391e20}, {0xf76ff4, 0x2391e08}, {0xf7495c, 0x2331e30})
	/home/runner/work/chezmoi/chezmoi/pkg/chezmoi/targetstateentry.go:81 +0x104
github.com/twpayne/chezmoi/v2/pkg/chezmoi.(*SourceState).Apply(0x2372180, {0xf782d0, 0x2391e20}, {0xf783fc, 0x238fb40}, {0xf76ff4, 0x2391e08}, {{0x0, 0x0}}, {{0x229ab10, ...}}, ...)
	/home/runner/work/chezmoi/chezmoi/pkg/chezmoi/sourcestate.go:681 +0xd5c
github.com/twpayne/chezmoi/v2/pkg/cmd.(*Config).applyArgs(0x2248000, {0xf757ac, 0x20280e0}, {0xf782d0, 0x2391e20}, {{0x0, 0x0}}, {0x2390800, 0x0, 0x1}, ...)
	/home/runner/work/chezmoi/chezmoi/pkg/cmd/config.go:619 +0x734
github.com/twpayne/chezmoi/v2/pkg/cmd.(*Config).runDumpCmd(0x2248000, 0x267e000, {0x2390800, 0x0, 0x1})
	/home/runner/work/chezmoi/chezmoi/pkg/cmd/dumpcmd.go:43 +0x118
github.com/spf13/cobra.(*Command).execute(0x267e000, {0x23907f8, 0x1, 0x1})
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.4.0/command.go:856 +0x6b0
github.com/spf13/cobra.(*Command).ExecuteC(0x2026580)
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.4.0/command.go:974 +0x414
github.com/spf13/cobra.(*Command).Execute(...)
	/home/runner/go/pkg/mod/github.com/spf13/cobra@v1.4.0/command.go:902
github.com/twpayne/chezmoi/v2/pkg/cmd.(*Config).execute(0x2248000, {0x2024020, 0x2, 0x2})
	/home/runner/work/chezmoi/chezmoi/pkg/cmd/config.go:1120 +0x78
github.com/twpayne/chezmoi/v2/pkg/cmd.runMain({{0xf6f338, 0x6}, {0xf76bb8, 0x28}, {0xf72d78, 0x14}, {0xf708d8, 0xa}}, {0x2024020, 0x2, ...})
	/home/runner/work/chezmoi/chezmoi/pkg/cmd/cmd.go:264 +0x148
github.com/twpayne/chezmoi/v2/pkg/cmd.Main({{0xf6f338, 0x6}, {0xf76bb8, 0x28}, {0xf72d78, 0x14}, {0xf708d8, 0xa}}, {0x2024020, 0x2, ...})
	/home/runner/work/chezmoi/chezmoi/pkg/cmd/cmd.go:128 +0x38
main.main()
	/home/runner/work/chezmoi/chezmoi/main.go:26 +0xd8
```

## Output of `chezmoi doctor`

<details>

```console
$ chezmoi doctor
RESULT    CHECK                MESSAGE
ok        version              v2.17.0, commit 7c70bff83d7c1dbe9ed3be486e3a6023e05c5426, built at 2022-05-26T00:05:04Z, built by goreleaser
ok        latest-version       v2.17.0
ok        os-arch              linux/arm (Raspbian GNU/Linux 11 (bullseye))
ok        uname                Linux cerberus 5.15.32-v7l+ #1538 SMP Thu Mar 31 19:39:41 BST 2022 armv7l GNU/Linux
ok        go-version           go1.18.2 (gc)
ok        executable           ~/bin/chezmoi
ok        upgrade-method       replace-executable
ok        config-file          ~/.config/chezmoi/chezmoi.yaml
ok        source-dir           ~/.local/share/chezmoi is a directory
ok        suspicious-entries   no suspicious entries
ok        working-tree         ~/.local/share/chezmoi is a directory
ok        dest-dir             ~ is a directory
ok        shell-command        found /usr/bin/zsh
ok        shell-args           /usr/bin/zsh
ok        cd-command           found /usr/bin/zsh
ok        cd-args              /usr/bin/zsh
ok        edit-command         found /bin/nano
ok        edit-args            /bin/nano
info      diff-command         not set
ok        umask                022
ok        git-command          found /usr/bin/git, version 2.30.2
warning   merge-command        bcompare not found in $PATH
info      age-command          age not found in $PATH
ok        gpg-command          found /usr/bin/gpg, version 2.2.27
info      pinentry-command     not set
info      1password-command    op not found in $PATH
info      bitwarden-command    bw not found in $PATH
info      gopass-command       gopass not found in $PATH
info      keepassxc-command    keepassxc-cli not found in $PATH
info      keeper-command       keeper not found in $PATH
info      keepassxc-db         not set
info      lastpass-command     lpass not found in $PATH
info      pass-command         pass not found in $PATH
info      vault-command        vault not found in $PATH
info      secret-command       not set
```

</details>

## Additional context

Presumably it is something in my dotfiles repo that triggers this. I have not made this public and is reluctant to do so as well. If you need I can try to "bisect" my chezmoi directory and try to find a small reproducer.

--- a/pkg/cmd/dumpcmd.go
+++ b/pkg/cmd/dumpcmd.go
@@ -23,7 +23,7 @@ func (c *Config) newDumpCmd() *cobra.Command {
 		ValidArgsFunction: c.targetValidArgs,
 		RunE:              c.runDumpCmd,
 		Annotations: map[string]string{
-			persistentStateMode:     persistentStateModeEmpty,
+			persistentStateMode:     persistentStateModeReadMockWrite,
 			requiresSourceDirectory: "true",
 		},
 	}

--- a/pkg/cmd/testdata/scripts/issue2092.txt
+++ b/pkg/cmd/testdata/scripts/issue2092.txt
@@ -1,0 +1,29 @@
+[windows] skip 'UNIX only'
+
+# test that chezmoi dump warns when the config file has not been generated
+chezmoi dump
+cmp stdout golden/dump.json
+stderr 'config file template has changed'
+
+# test that chezmoi dump does not return an error when the config file template has been modified
+chezmoi init
+edit $CHEZMOICONFIGDIR${/}chezmoi.toml
+chezmoi dump
+cmp stdout golden/dump.json
+! stderr .
+
+-- golden/dump.json --
+{
+  ".file": {
+    "type": "file",
+    "name": ".file",
+    "contents": "# contents of .file\n",
+    "perm": 420
+  }
+}
+-- home/user/.file --
+# contents of .file
+-- home/user/.local/share/chezmoi/.chezmoi.toml.tmpl --
+[data]
+-- home/user/.local/share/chezmoi/dot_file --
+# contents of .file

